# Exploring Ebay Car Sales Data

In this project, we will clean and analyze the dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle. 

In [1]:
import numpy as np
import pandas as pd
autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
print("*********************************")
print("General information about the dataset")
print("*********************************")
autos.info()
print("*********************************")
print("         Top 5 rows              ")
print("*********************************")
print(autos.head())

*********************************
General information about the dataset
*********************************
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 

It's been found that there are 50000 rows and 20 columns in the dataset. 15 columns are of object type and 5 columns are of int64 type. Also null values is observed in the dataset.

In [4]:
#List of columns
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

Its been observed that the column names are of camelcase instead of python's prefered snakecase. So we will convert the column names to snakecase.

In [5]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Let's explore the data further to determine the cleaning tasks that need's to be performed.

In [6]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-16 21:50:53  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

### Exploring the Odometer and Price Columns

From the observation made above we have decided to,

1. Drop any columns that have mostly one value.
2. Convert numeric data stored as text.
3. Investigate any columns if required.

In [7]:
#Drop columns
autos = autos.drop(['nr_of_pictures','seller','offer_type'],axis = 1)

In [8]:
# Removing non-numeric characters and converting to numeric type
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)
autos["odometer"] = autos["odometer"].str.replace("km","").str.replace(",","").astype(int)

In [9]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   price               50000 non-null  int32 
 3   abtest              50000 non-null  object
 4   vehicle_type        44905 non-null  object
 5   registration_year   50000 non-null  int64 
 6   gearbox             47320 non-null  object
 7   power_ps            50000 non-null  int64 
 8   model               47242 non-null  object
 9   odometer            50000 non-null  int32 
 10  registration_month  50000 non-null  int64 
 11  fuel_type           45518 non-null  object
 12  brand               50000 non-null  object
 13  unrepaired_damage   40171 non-null  object
 14  ad_created          50000 non-null  object
 15  postal_code         50000 non-null  int64 
 16  last_seen           50

In [10]:
# Since we have removed the "km" character in the odometer column,
# the column name is changed inorder to understand that the values
#are in km

autos.rename({"odometer" : "odometer_km"}, axis = 1, inplace = True)

Let's explore the price and odometer_km columns further, especially looking for data that doesn't look right.

In [11]:
autos["odometer_km"].unique().shape

(13,)

In [12]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [13]:
autos["odometer_km"].value_counts().sort_index()

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

**Odometer :** All the values seems to be in a normal range where the minimum value is 5000 kms and the maximum value is 150000 kms. Majority of the used cars listed with an odmoter reading of 150000 kms.

There are no outliers we need to get rid of in this column.

In [14]:
autos["price"].unique().shape

(2357,)

In [15]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [16]:
autos["price"].value_counts().sort_index()

0           1421
1            156
2              3
3              1
5              2
            ... 
10000000       1
11111111       2
12345678       3
27322222       1
99999999       1
Name: price, Length: 2357, dtype: int64

**Price :** The lowest values is 0 Euro and the maximum value seems to be 100 million euros which is unusual. Also there are many unusual values in the dataset like 10000000,12345678, 99999999. Many cars used to be very old or non-functional so the owner would have posted it for 0 euros (free).   

There are outliers that we need to get rid of in this column.

In [17]:
#List of cars above 200000 euros.
autos[autos["price"] > 200000].sort_values("price")      

date_crawled                                               name  \
37840  2016-03-21 10:50:12                                        Porsche_997   
38299  2016-03-28 22:25:25                                Glas_BMW_mit_Wasser   
47337  2016-04-05 10:25:38                                    BMW_Z8_roadster   
12682  2016-03-28 22:48:01  Porsche_GT3_RS__PCCB__Lift___grosser_Exklusiv_...   
35923  2016-04-03 07:56:23  Porsche_911_Targa_Exclusive_Edition__1_von_15_...   
34723  2016-03-23 16:37:29  Porsche_Porsche_911/930_Turbo_3.0__deutsche_Au...   
14715  2016-03-30 08:37:24                 Rolls_Royce_Phantom_Drophead_Coupe   
36818  2016-03-27 18:37:37                                        Porsche_991   
37585  2016-03-29 11:38:54                                Volkswagen_Jetta_GT   
43049  2016-03-21 19:53:52                                      2_VW_Busse_T3   
514    2016-03-17 09:53:08                   Ford_Focus_Turnier_1.6_16V_Style   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
47634  2016-04-04 21:25:21                                        Ferrari_FXX   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   

          price   abtest vehicle_type  registration_year    gearbox  power_ps  \
37840    220000     test        coupe               2008    manuell       415   
38299    250000     test          NaN               2015        NaN         0   
47337    259000     test       cabrio               2001    manuell       400   
12682    265000  control        coupe               2016  automatik       500   
35923    295000     test       cabrio               2015  automatik       400   
34723    299000     test        coupe               1977    manuell       260   
14715    345000  control       cabrio               2012  automatik       460   
36818    350000  control        coupe               2016    manuell       500   
37585    999990     test    limousine               1985    manuell       111   
43049    999999     test          bus               1981    manuell        70   
514      999999     test        kombi               2009    manuell       101   
22947   1234566  control        kombi               1999  automatik       190   
7814    1300000  control        coupe               1992        NaN         0   
47634   3890000     test        coupe               2006        NaN       799   
11137  10000000  control        coupe               1960    manuell       368   
24384  11111111     test          NaN               1995        NaN         0   
2897   11111111     test    limousine               1973    manuell        48   
27371  12345678  control          NaN               2017        NaN        95   
39377  12345678  control          NaN               2018    manuell        95   
47598  12345678  control    limousine               2001    manuell       101   
42221  27322222  control    limousine               2014    manuell       163   
39705  99999999  control    limousine               1999  automatik       224   

             model  odometer_km  registration_month fuel_type           brand  \
37840          911        30000                   7    benzin         porsche   
38299      x_reihe         5000                   0       NaN             bmw   
47337      z_reihe        20000            

Since luxury cars cost more than 200000 euros, we shall delete the car details whose price is more than 500000 euros.

In [18]:
autos = autos[autos["price"] < 500000]
autos["price"].describe()

count     49986.000000
mean       5721.525167
std        8983.617820
min           0.000000
25%        1100.000000
50%        2950.000000
75%        7200.000000
max      350000.000000
Name: price, dtype: float64

In [19]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 49986 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        49986 non-null  object
 1   name                49986 non-null  object
 2   price               49986 non-null  int32 
 3   abtest              49986 non-null  object
 4   vehicle_type        44894 non-null  object
 5   registration_year   49986 non-null  int64 
 6   gearbox             47310 non-null  object
 7   power_ps            49986 non-null  int64 
 8   model               47233 non-null  object
 9   odometer_km         49986 non-null  int32 
 10  registration_month  49986 non-null  int64 
 11  fuel_type           45509 non-null  object
 12  brand               49986 non-null  object
 13  unrepaired_damage   40163 non-null  object
 14  ad_created          49986 non-null  object
 15  postal_code         49986 non-null  int64 
 16  last_seen           49

## Exploring date columns

It's been noted that certain columns that represent date values are of object & int type.

Let's examine the following columns - date_crawled, last_seen, ad_created.

1. Format followed 
2. Range of values

In [20]:
autos[["date_crawled", "last_seen", "ad_created"]].describe()

date_crawled            last_seen           ad_created
count                 49986                49986                49986
unique                48200                39472                   76
top     2016-03-05 16:57:05  2016-04-07 06:17:27  2016-04-03 00:00:00
freq                      3                    8                 1946

In [21]:
autos[["date_crawled", "last_seen", "ad_created"]].head()

date_crawled            last_seen           ad_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00

In [22]:
# Distribution of date_crawled
autos["date_crawled"].str[:10].value_counts(dropna=False).sort_index()

2016-03-05    1269
2016-03-06     697
2016-03-07    1798
2016-03-08    1663
2016-03-09    1660
2016-03-10    1606
2016-03-11    1624
2016-03-12    1838
2016-03-13     778
2016-03-14    1831
2016-03-15    1699
2016-03-16    1475
2016-03-17    1575
2016-03-18     653
2016-03-19    1745
2016-03-20    1891
2016-03-21    1874
2016-03-22    1645
2016-03-23    1619
2016-03-24    1455
2016-03-25    1587
2016-03-26    1624
2016-03-27    1552
2016-03-28    1742
2016-03-29    1707
2016-03-30    1681
2016-03-31    1595
2016-04-01    1690
2016-04-02    1770
2016-04-03    1934
2016-04-04    1824
2016-04-05     655
2016-04-06     159
2016-04-07      71
Name: date_crawled, dtype: int64

In [23]:
# Distribution of last_seen
autos["last_seen"].str[:10].value_counts(dropna=False).sort_index()

2016-03-05       54
2016-03-06      221
2016-03-07      268
2016-03-08      379
2016-03-09      492
2016-03-10      538
2016-03-11      626
2016-03-12     1190
2016-03-13      449
2016-03-14      640
2016-03-15      794
2016-03-16      822
2016-03-17     1396
2016-03-18      371
2016-03-19      787
2016-03-20     1035
2016-03-21     1036
2016-03-22     1079
2016-03-23      929
2016-03-24      978
2016-03-25      960
2016-03-26      848
2016-03-27      801
2016-03-28     1042
2016-03-29     1116
2016-03-30     1242
2016-03-31     1191
2016-04-01     1155
2016-04-02     1244
2016-04-03     1268
2016-04-04     1231
2016-04-05     6212
2016-04-06    11046
2016-04-07     6546
Name: last_seen, dtype: int64

The ads that have been seen during early dates must have been removed since the car could have been sold out.

In [24]:
# Distribution of ad_created
autos["ad_created"].str[:10].value_counts(dropna=False).sort_index()

2015-06-11       1
2015-08-10       1
2015-09-09       1
2015-11-10       1
2015-12-05       1
              ... 
2016-04-03    1946
2016-04-04    1842
2016-04-05     592
2016-04-06     163
2016-04-07      64
Name: ad_created, Length: 76, dtype: int64

It's been observed that the all the 3 columns follow the same formate were the first 10 characters represents the date. The distribution of date_crawled & last_seen lies between March 05,2016 to April 07,2016. However there are a few dates in the ad_created column were ads are created in the year prior to that timeframe. 

In [25]:
autos[["registration_year", "registration_month"]].describe()

registration_year  registration_month
count       49986.000000        49986.000000
mean         2005.075721            5.723723
std           105.727161            3.711839
min          1000.000000            0.000000
25%          1999.000000            3.000000
50%          2003.000000            6.000000
75%          2008.000000            9.000000
max          9999.000000           12.000000

It's been observed that there are many inaccurate data. For instance, the minimum value in the registraction_year column is 1000 & the maximum is 9999. Its difficult to find the earliest valid year.

### Removing inaccurate registration year

Initially, lets count the number of listing with the cars that fall outside 1900 - 2016 intreval.

In [26]:
print(autos[(autos["registration_year"] > 2016) | 
                         (autos["registration_year"] < 1900)]
                          ["registration_year"].value_counts().
                          sort_index())

1000       1
1001       1
1111       1
1500       1
1800       2
2017    1452
2018     491
2019       3
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       2
9996       1
9999       4
Name: registration_year, dtype: int64


The above results seems to be satisfiable. However, we do have many rows with registration year as 2017 or 2018. Perhaps these are 2017 or 2018 model year cars in which the person selling, filled in the registration_year field incorrectly. Let's keep these cars as to now skew results too much, but note that these values still exist in our data.

In [27]:
autos = autos[(autos["registration_year"] > 1900) & (autos["registration_year"] <= 2018)]
autos[["registration_year", "registration_month"]].describe()

registration_year  registration_month
count       49959.000000        49959.000000
mean         2003.367862            5.725595
std             7.688527            3.711096
min          1910.000000            0.000000
25%          1999.000000            3.000000
50%          2003.000000            6.000000
75%          2008.000000            9.000000
max          2018.000000           12.000000

Finally, we have removed the rows with inaccurate registration_year from our dataset.

## Determine the average price of the brands which contibutes major market share.

In [28]:
market_share = autos["brand"].value_counts(normalize=True)
print(market_share)

volkswagen        0.213755
opel              0.109230
bmw               0.108629
mercedes_benz     0.094698
audi              0.085730
ford              0.069597
renault           0.048119
peugeot           0.029144
fiat              0.026161
seat              0.018815
skoda             0.015713
mazda             0.015152
nissan            0.015092
smart             0.014032
citroen           0.013991
toyota            0.012350
sonstige_autos    0.010769
hyundai           0.009768
volvo             0.009127
mini              0.008487
mitsubishi        0.008087
honda             0.007967
kia               0.007126
alfa_romeo        0.006565
porsche           0.005885
suzuki            0.005865
chevrolet         0.005665
chrysler          0.003623
dacia             0.002582
daihatsu          0.002562
jeep              0.002182
subaru            0.002162
land_rover        0.001982
daewoo            0.001581
saab              0.001581
jaguar            0.001541
trabant           0.001541
r

Since Volkswagen, BMW, OPEL and Mercedez Benz contributes over 50% of the market share. We will explore the average price of the brands having market share above 0.05.

In [29]:
# Brands with market share above 0.05
top_brands = market_share[market_share > 0.05]
print(top_brands)

volkswagen       0.213755
opel             0.109230
bmw              0.108629
mercedes_benz    0.094698
audi             0.085730
ford             0.069597
Name: brand, dtype: float64


## Exploring price by brand

In [30]:
average_price_dict = {}
for item in top_brands.index:
    brand = autos[autos["brand"] == item]
    avg_price = brand["price"].mean()
    average_price_dict[item] = int(avg_price)
print(average_price_dict)

{'volkswagen': 5159, 'opel': 2842, 'bmw': 8028, 'mercedes_benz': 8380, 'audi': 8965, 'ford': 3626}


We observed that, there is a distinct price gap.

1. Audi, BMW and Mercedes Benz are more expensive
2. Ford and Opel are less expensive
3. Volkswagen is in between

## Exploring mileage by brand

In [31]:
average_mileage_dict = {}
for item in top_brands.index:
    brand = autos[autos["brand"] == item]
    avg_mileage = brand["odometer_km"].mean()
    average_mileage_dict[item] = int(avg_mileage)
print(average_mileage_dict)

{'volkswagen': 129006, 'opel': 129361, 'bmw': 132540, 'mercedes_benz': 130933, 'audi': 129643, 'ford': 124153}


In [32]:
mean_price = pd.Series(average_price_dict)
mean_mileage = pd.Series(average_mileage_dict)
top_autos_stats = pd.DataFrame(mean_price, columns = ["MEAN_PRICE"])
top_autos_stats["MEAN_MILEAGE"] = pd.DataFrame(mean_mileage)
top_autos_stats.sort_values(by=['MEAN_MILEAGE'])

MEAN_PRICE  MEAN_MILEAGE
ford                 3626        124153
volkswagen           5159        129006
opel                 2842        129361
audi                 8965        129643
mercedes_benz        8380        130933
bmw                  8028        132540

It can be deduced from this that the range of car mileages (odometer_km) does not vary as much as the prices do by brand. However, there is a slight trend to the less expensive cars having lower mileage, with the more expensive vehicles having higher mileage.